File: Data Structures & Algorithms/validate-parentheses/submission-15.py
class Solution:
    def isValid(self, s: str) -> bool:
        closing_brackets = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        stack = []

        for c in s:
            if c in closing_brackets:
                if stack and stack[-1] == closing_brackets[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return not stack

File: Data Structures & Algorithms/minimum-stack/submission-15.java
class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public MinStack() {
        this.stack = new Stack(); 
        this.minStack = new Stack(); 
    }
    
    public void push(int val) {
        stack.push(val);
        if (minStack.isEmpty()){
            minStack.push(0);
        }else{
            if (val < stack.get(minStack.peek())){
                minStack.push(stack.size()-1);
            }
        }
    }
    
    public void pop() {
        stack.pop();
        if (stack.size() == minStack.peek()){
            minStack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return stack.get(minStack.peek());
    }
}
